src/dune_engine/build_system.mli

```diff
@@ -11,7 +11,7 @@ module Error : sig
   (** Errors when building a target *)
   type t
 
-  val info : t -> User_message.t * Dep_path.Entry.t list option * Path.t option
+  val info : t -> User_message.t * Path.t option
```

Comment:
Continuing the conversation from slack: (cc @cwong-ocaml)

> if instead of a structured value, you get a Pp.t list  (= List.map ~f:Dep_path.Entry.pp <current list>), would that be sufficient for the RPC?

it wouldn't  be sufficient lsp. In lsp, every single error needs to belong to a source file (the path that we currently provide is the  directory where the rule is executed - but it is irrelevant). I intended to extract this source file from the dep_path stack, but now I need to use this new mechanism and I'm not quite sure it's enough. The information I need is basically already pushed in build_file_impl, but I don't know how to extract it. Is there a way to access the memo stack frame for the error?

Reply:
You could do it the same way we recover locations for "no rule found for ..." errors. For these, we walk up the memo stack and look for stack frames that correspond to a call to "execute_rule <rule>":

```
        List.find_map stack ~f:(fun frame ->
            match
              Memo.Stack_frame.as_instance_of frame ~of_:execute_rule_memo
            with
            | Some r -> Some (Rule.loc r)
            | None -> None)
```

For your purpose, you could look for `build_file_memo`, the argument would be the file that was requested to build.

However, I'm not sure that's going to be much help because you are not going to find a source file this way. What's you'll get is the cmx or cmo. That is true with the new code, but was also true with the old `Dep_path` module.

Reply:
> However, I'm not sure that's going to be much help because you are not going to find a source file this way. What you'll get is the cmx or cmo. That is true with the new code, but was also true with the old Dep_path module.

I think this can be addressed using our standard trick for testing if a path is in the source. We trim the build context, and check if the result exists in Source_tree.

Reply:
BTW, I remember having a similar discussion with @cwong-ocaml. One idea would be look up the stack for the first rule that has dependencies in the source tree. Such source files are likely to be the one you care about.